scipy/_lib/_array_api.py

```diff
@@ -196,19 +196,26 @@ def is_torch(xp):
     return xp.__name__ == 'scipy._lib.array_api_compat.array_api_compat.torch'
 
 
-def _strict_check(actual, desired, check_xp=True, check_dtype=True, check_shape=True):
-    if check_xp:
+def _strict_check(actual, desired, xp,
+                  check_namespace=True, check_dtype=True, check_shape=True):
+    if check_namespace:
         _assert_matching_namespace(actual, desired)
-    xp = array_namespace(actual)
+
     actual = xp.asarray(actual)
     desired = xp.asarray(desired)
+
     if check_dtype:
         assert_(actual.dtype == desired.dtype,
-                f"Desired dtype: {desired.dtype}, actual dtype: {actual.dtype}")
+                "dtypes do not match.\n"
+                f"Actual: {actual.dtype}\n"
+                f"Desired: {desired.dtype}")
+
     if check_shape:
-        msg = (f"Array shapes are not equal: actual shape = {actual.shape}, "
-               f"desired shape = {desired.shape}")
-        assert_(actual.shape == desired.shape, msg)
+        assert_(actual.shape == desired.shape,
+                "Shapes do not match.\n"
+                f"Actual: {actual.shape}\n"
+                f"Desired: {desired.shape}")
+        _check_scalar(actual, desired, xp)
```

Comment:
I originally suggested that `_check_scalar` would be part of `check_namespace`/`check_xp`, but I think it's more appropriate for it to be part of the shape check (or something separate altogether).